spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpNioConnection.java

```diff
@@ -662,6 +662,8 @@ protected synchronized void doWrite(ByteBuffer buffer) throws IOException {
 				TcpNioConnection.this.socketChannel.write(buffer);
 				remaining = buffer.remaining();
 			}
+			selector.close();
+			selector = null;
```

Comment:
And why do we need to close selector after each write?
You said yourself in the PR description:

> Closing/destroying `ChannelOutputStream` object does not close the selector 

I don't see that you close a `ChannelOutputStream` in this write operation, so I believe it is done somehow externally. Therefore that `ChannelOutputStream.close()` is called from there. 
So, it still looks suspicious to always close a selector in the end of write, but not once when `ChannelOutputStream` is closed.

What do I miss, please?

Reply:
OK I see your point it is better to move ```selector.close()``` to ```ChannelOutputStream.close()```
And I would like to ask you if you feel comfortable with:
```ChannelOutputStream.close()``` eventually calls ```doClose()``` that eventually calls ```super.close()``` that has the following code:
`for (TcpSender sender : this.senders) {
				sender.removeDeadConnection(this);`

This ```removeDeadConnection``` does nothing it has empty implementation, I feel (maybe wrong) that it may cause Out Of Memory in one of the maps in my case in ```TcpSendingMessageHandler``` in map ```connections```.

Reply:
I think that is not related to what we are talking here about.
I guess @garyrussell is looking into that via #3701 .

Let's concentrate here on the proper selector closing!